Allow write access to offsets of readonly ArrayAccess properties

pilif opened the pull request:
In case a readonly property is an object which implements ArrayAccess, offet accesses to said property are turned into method calls on PHP's side and are allowed even on read-only properties.

this fixes phpstan/phpstan#8929

pilif committed:
Allow write access to offsets of readonly ArrayAccess properties

In case a readonly property is an object which implements ArrayAccess,
offet accesses to said property are turned into method calls on PHP's
side and are allowed even on read-only properties.

this fixes phpstan/phpstan#8929

pilif commented:
one sec - will fix to make this work properly with old PHP versions

pilif commented:
ok. much better.

I don't think the failing rector integration test is related to this at all

ondrejmirtes commented on src/Rules/Properties/ReadOnlyPropertyAssignRule.php:
```diff
@@ -89,6 +91,10 @@ public function processNode(Node $node, Scope $scope): array
 			}
 
+			if ((new ObjectType(ArrayAccess::class))->isSuperTypeOf($propertyReflection->getNativeType())->yes()) {
+				continue;
```
This is a wrong fix. You can't say that in general.

If I have `private readonly ArrayAcess $a` and I do `$this->a = new Foo()` twice, PHP does not allow that.

pilif replied:
🤦 of course - you're right.

I have pushed a patch to expand the test to test for your case and to also check that we're dealing with a `SetOffsetValueTypeExpr`

pilif committed:
fix checks for phpstan/phpstan#8929

of course just checking for ArrayAccess isn't enough - it's only valid
for index assignment operations 🤦

herndlm commented:
Interesting, wasn't aware of this. Looks like my https://github.com/phpstan/phpstan-src/pull/3817 is doing a very similar thing, but we mention different issues that are fixed. Those might be duplicates, but can't check the details right now..

ondrejmirtes commented:
Thank you. I liked https://github.com/phpstan/phpstan-src/pull/3817 slightly more.
